File: kotlin/hydroponics-host/hydroponics-host/src/main/kotlin/io/greenglass/hydroponics/application/models/installation/InstallationNodeModel.kt
package io.greenglass.hydroponics.application.models.installation
import io.greenglass.hydroponics.application.models.node.METRIC
import io.greenglass.hydroponics.application.models.node.invoke
import io.greenglass.sparkplug.models.Metric
import org.neo4j.graphdb.Direction
import org.neo4j.graphdb.Node

class InstallationNodeModel(
    val nodeId : String,
    val implName : String,
    val nodeType : String,
    val nodeName : String,
    val nodeDescription : String,
    val nodeImage : String,
    val metrics  : List<Metric>

) {
    constructor(modeId: String, implName : String, nodeType : Node) : this(
        modeId,
        implName,
        nodeType.getProperty("type") as String,
        nodeType.getProperty("name") as String,
        nodeType.getProperty("description") as String,
        nodeType.getProperty("image") as String,
        nodeType.getRelationships(Direction.OUTGOING, Metric.METRIC)
            .map { r -> r.endNode }
            .map { n -> Metric.Companion.invoke(n) }
        )
}

File: flutter/hydroponics_framework/lib/datamodels/installation/installation_node_model.dart
import 'package:logger/logger.dart';

import '../implementation/implementation_node_model.dart';
import '../nodes/metric.dart';

class InstallationNodeModel {
  final String nodeId;
  final String implName;
  final String nodeType;
  final String nodeName;
  final String nodeDescription;
  final String nodeImage;
  final List<Metric> metrics;

  InstallationNodeModel(this.nodeId, this.implName, this.nodeType,
      this.nodeName, this.nodeDescription, this.nodeImage, this.metrics);

  static InstallationNodeModel fromJson(Map<String, dynamic> json) {
    return InstallationNodeModel(
        json["nodeId"],
        json["implName"],
        json["nodeType"],
        json["nodeName"],
        json["nodeDescription"],
        json["nodeImage"],
        List.from(json["metrics"]).map((m) => Metric.fromJson(m)).toList());
  }
}
